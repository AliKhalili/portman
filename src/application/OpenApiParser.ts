import SwaggerParser from '@apidevtools/swagger-parser'
import { OpenAPIV3 } from 'openapi-types'
import path from 'path'
import { MappedOperation } from '../lib/oas/MappedOperation'

export interface OpenApiParserConfig {
  inputFile: string
}

export class OpenApiParser {
  public oas: OpenAPIV3.Document
  public mappedOperations: MappedOperation[]

  async convert(options: OpenApiParserConfig): Promise<OpenAPIV3.Document> {
    const inputFile = path.resolve(options.inputFile)
    // Dereference the spec so all entities have encapsulated schemas
    const api = await SwaggerParser.dereference(inputFile)
    const oasDoc = (await SwaggerParser.bundle(api, {
      parse: {
        json: false, // Disable the JSON parser
        yaml: {
          allowEmpty: false // Don't allow empty YAML files
        }
      },
      dereference: {
        circular: false // Don't allow circular $refs
      },
      validate: {
        spec: false // Don't validate against the Swagger spec
      }
    })) as OpenAPIV3.Document

    this.oas = oasDoc
    this.mappedOperations = this.pathsToOperations()
    return oasDoc
  }

  pathsToOperations = (): MappedOperation[] => {
    const paths = this.oas.paths
    const mappedOperations = Object.entries(paths)
      .filter(([_path, operations]) => !!operations)
      .map(([path, operations]) => {
        return (
          operations &&
          Object.entries(operations).map(([method, operation]): MappedOperation => {
            return new MappedOperation(path, method, operation as OpenAPIV3.OperationObject)
          })
        )
      })
      .reduce<MappedOperation[]>((acc, resource) => {
        resource && resource.map(item => acc.push(item))
        return acc
      }, [])

    this.mappedOperations = mappedOperations
    return this.mappedOperations
  }

  public getOperationById(operationId: string): MappedOperation | null {
    return this.mappedOperations.find(({ id }) => id === operationId) || null
  }

  public getOperationByPath(path: string): MappedOperation | null {
    return this.mappedOperations.find(({ pathRef }) => pathRef === path) || null
  }
}
