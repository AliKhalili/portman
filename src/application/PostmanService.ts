import axios, { AxiosRequestConfig } from 'axios'
import chalk from 'chalk'
import { CollectionDefinition } from 'postman-collection'

export class PostmanService {
  private baseUrl: string
  private apiKey: string

  constructor() {
    this.baseUrl = 'https://api.getpostman.com'
    this.apiKey = `${process.env.POSTMAN_API_KEY}`
  }

  async updateCollection(collection: CollectionDefinition, uuid: string): Promise<string> {
    const data = JSON.stringify({
      collection: collection
    })

    const config = {
      method: 'put',
      url: `${this.baseUrl}/collections/${uuid}`,
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.apiKey
      },
      data: data
    } as AxiosRequestConfig

    try {
      const res = await axios(config)
      const data = res.data
      console.log('Upload to Postman Success:', data)
      return JSON.stringify(data, null, 2)
    } catch (error) {
      console.log(
        chalk.red(`=================================================================
      `)
      )
      console.log(
        chalk.red(`Upload to Postman Failed'
      `)
      )
      console.log(error?.response?.data)
      console.log(`\n`)
      console.log(
        chalk.red(`=================================================================
      `)
      )
      return error.toString()
    }
  }
}
