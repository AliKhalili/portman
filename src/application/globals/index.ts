export * from './injectEnvVariables'
export * from './orderCollectionRequests'
export * from './overwriteCollectionKeyValues'
export * from './overwriteCollectionValues'
export * from './overwriteSecurity'
export * from './writeCollectionPreRequestScripts'
export * from './writeRawReplacements'
