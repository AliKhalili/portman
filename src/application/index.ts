export * from './DownloadService'
export * from './OpenApiParser'
export * from './OpenApiToPostmanService'
export * from './PostmanService'
