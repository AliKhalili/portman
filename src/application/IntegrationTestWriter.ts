import { Collection, Item, ItemGroup } from 'postman-collection'
import { IntegrationTestConfig } from 'src/types'
import { TestSuite, VariationWriter } from './'

export class IntegrationTestWriter {
  public testSuite: TestSuite
  integrationTestFolder: ItemGroup<Item>
  integationTestCollection: Collection

  constructor({
    testSuite,
    integrationTestFolderName
  }: {
    testSuite: TestSuite
    integrationTestFolderName: string
  }) {
    this.testSuite = testSuite
    this.integrationTestFolder = new ItemGroup<Item>({
      name: integrationTestFolderName
    })
    this.integationTestCollection = new Collection()
  }

  public add(integrationTest: IntegrationTestConfig): void {
    const { testSuite } = this
    const { name, operations } = integrationTest

    const variationWriter = new VariationWriter({
      testSuite: this.testSuite,
      variationFolderName: name
    })

    operations.map(({ openApiOperationId, variations }) => {
      const pmOperations = testSuite.postmanParser.getOperationsByIds([openApiOperationId])
      const pmOperation = pmOperations[0]
      const oaOperation = testSuite.oasParser.getOperationByPath(pmOperation.pathRef)

      variations.map(variation => {
        const variationName = variation.name

        const operationVariation = pmOperation.clone({
          newId: `${variationName}-${Math.random()}`,
          name: variationName
        })

        variationWriter.injectVariations(operationVariation, oaOperation, variation)
        variationWriter.addToFolder(operationVariation, variationWriter.variationFolder)
      })
    })

    this.integationTestCollection = variationWriter.mergeToCollection(this.integationTestCollection)
  }

  public mergeToCollection(collection: Collection): Collection {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    this.integationTestCollection.items.map((item: any) => {
      this.integrationTestFolder.items.add(item)
    })

    collection.items.add(this.integrationTestFolder)
    return collection
  }
}
