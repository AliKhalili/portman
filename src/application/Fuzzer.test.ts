import { TestSuite, VariationWriter } from '../application'
import { PostmanParser } from '../postman'
import { OpenApiParser } from '../oas'
import fs from 'fs-extra'
import { getConfig } from '../lib'
import { Collection } from 'postman-collection'
import { Fuzzer } from './Fuzzer'
import { FuzzingSchemaItems, PortmanFuzzTypes } from '../types'

describe('Fuzzer', () => {
  let postmanParser: PostmanParser
  let oasParser: OpenApiParser
  let testSuite: TestSuite
  let variationWriter: VariationWriter
  let variationTests
  let variationTest
  let variationMeta
  let fuzzer: Fuzzer
  let pmOps
  let pmOp
  let oaOp

  const postmanJson = '__tests__/fixtures/crm.postman.json'
  const oasYml = '__tests__/fixtures/crm.yml'
  const portmanConfigFile = '__tests__/fixtures/portman-fuzzing.crm.json'

  beforeEach(async () => {
    oasParser = new OpenApiParser()
    await oasParser.convert({ inputFile: oasYml })
    const postmanObj = JSON.parse(fs.readFileSync(postmanJson).toString())
    const config = await getConfig(portmanConfigFile)
    postmanParser = new PostmanParser({
      collection: new Collection(postmanObj),
      oasParser: oasParser
    })
    testSuite = new TestSuite({ oasParser, postmanParser, config })
    variationWriter = new VariationWriter({
      testSuite: testSuite,
      variationFolderName: 'Variation Tests'
    })

    fuzzer = new Fuzzer({ testSuite: testSuite, variationWriter })

    variationTests = config?.tests?.variationTests
    variationTest = variationTests[0]
    variationMeta = variationTest.variationMeta

    pmOps = testSuite.getOperationsFromSetting(variationTest)
    pmOp = pmOps[11] // POST Leads operation
    oaOp = oasParser.getOperationByPath(pmOp.pathRef)
  })

  it('should not fuzz', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: [],
      minimumNumberFields: [],
      maximumNumberFields: [],
      minLengthFields: [],
      maxLengthFields: []
    } as FuzzingSchemaItems

    fuzzer.injectFuzzRequiredVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[0]
    expect(result).toBeUndefined()
  })

  it('should fuzz the required prop of the request body', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: ['name'],
      minimumNumberFields: [],
      maximumNumberFields: [],
      minLengthFields: [],
      maxLengthFields: []
    } as FuzzingSchemaItems

    fuzzer.injectFuzzRequiredVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[0]
    expect(result.item.request?.body?.raw).toMatchSnapshot()
  })

  it('should fuzz the multiple required props of the request body', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: ['name', 'company_name'],
      minimumNumberFields: [],
      maximumNumberFields: [],
      minLengthFields: [],
      maxLengthFields: []
    } as FuzzingSchemaItems

    fuzzer.injectFuzzRequiredVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[1]
    expect(result.item.request?.body?.raw).toMatchSnapshot()
  })

  it('should fuzz the value of a prop of the request body below the defined minimum', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: [],
      minimumNumberFields: [{ path: 'monetary_amount', field: 'monetary_amount', value: 100 }],
      maximumNumberFields: [],
      minLengthFields: [],
      maxLengthFields: []
    } as FuzzingSchemaItems

    fuzzer.injectFuzzMinimumVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[0]
    expect(result.item.request?.body?.raw).toMatchSnapshot()
  })

  it('should fuzz the value of a prop of the request body above the defined maximum', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: [],
      minimumNumberFields: [],
      maximumNumberFields: [{ path: 'monetary_amount', field: 'monetary_amount', value: 300 }],
      minLengthFields: [],
      maxLengthFields: []
    } as FuzzingSchemaItems

    fuzzer.injectFuzzMaximumVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[0]
    expect(result.item.request?.body?.raw).toMatchSnapshot()
  })

  it('should fuzz the length of a prop of the request body above the defined minimum', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: [],
      minimumNumberFields: [],
      maximumNumberFields: [],
      minLengthFields: [{ path: 'first_name', field: 'first_name', value: 11 }],
      maxLengthFields: []
    } as FuzzingSchemaItems

    fuzzer.injectFuzzMinLengthVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[0]
    expect(result.item.request?.body?.raw).toMatchSnapshot()
  })

  it('should fuzz the length of a prop of the request body above the defined maximum', async () => {
    const fuzzItems = {
      fuzzType: PortmanFuzzTypes.requestBody,
      requiredFields: [],
      minimumNumberFields: [],
      maximumNumberFields: [],
      minLengthFields: [],
      maxLengthFields: [{ path: 'first_name', field: 'first_name', value: 10 }]
    } as FuzzingSchemaItems

    fuzzer.injectFuzzMaxLengthVariation(pmOp, oaOp, variationTest, variationMeta, fuzzItems)

    const result = fuzzer.fuzzVariations[0]
    expect(result.item.request?.body?.raw).toMatchSnapshot()
  })
})
