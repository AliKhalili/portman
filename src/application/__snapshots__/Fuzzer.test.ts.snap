// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fuzzer should fuzz the length of a prop of the request body above the defined maximum 1`] = `
"{
    \\"name\\": \\"Elon Musk\\",
    \\"company_name\\": \\"Spacex\\",
    \\"owner_id\\": \\"54321\\",
    \\"company_id\\": \\"2\\",
    \\"contact_id\\": \\"2\\",
    \\"first_name\\": \\"ElonEEEE\\",
    \\"last_name\\": \\"Musk\\",
    \\"description\\": \\"A thinker\\",
    \\"prefix\\": \\"Sir\\",
    \\"title\\": \\"CEO\\",
    \\"status\\": \\"New\\",
    \\"monetary_amount\\": 75000,
    \\"currency\\": \\"USD\\",
    \\"fax\\": \\"+12129876543\\",
    \\"websites\\": [
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"addresses\\": [
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        },
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        }
    ],
    \\"social_links\\": [
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        },
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        }
    ],
    \\"phone_numbers\\": [
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"emails\\": [
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"custom_fields\\": [
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        },
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        }
    ],
    \\"tags\\": [
        \\"New\\"
    ]
}"
`;

exports[`Fuzzer should fuzz the length of a prop of the request body above the defined minimum 1`] = `
"{
    \\"name\\": \\"Elon Musk\\",
    \\"company_name\\": \\"Spacex\\",
    \\"owner_id\\": \\"54321\\",
    \\"company_id\\": \\"2\\",
    \\"contact_id\\": \\"2\\",
    \\"first_name\\": \\"Elon\\",
    \\"last_name\\": \\"Musk\\",
    \\"description\\": \\"A thinker\\",
    \\"prefix\\": \\"Sir\\",
    \\"title\\": \\"CEO\\",
    \\"status\\": \\"New\\",
    \\"monetary_amount\\": 75000,
    \\"currency\\": \\"USD\\",
    \\"fax\\": \\"+12129876543\\",
    \\"websites\\": [
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"addresses\\": [
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        },
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        }
    ],
    \\"social_links\\": [
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        },
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        }
    ],
    \\"phone_numbers\\": [
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"emails\\": [
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"custom_fields\\": [
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        },
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        }
    ],
    \\"tags\\": [
        \\"New\\"
    ]
}"
`;

exports[`Fuzzer should fuzz the multiple required props of the request body 1`] = `
"{
    \\"name\\": \\"Elon Musk\\",
    \\"owner_id\\": \\"54321\\",
    \\"company_id\\": \\"2\\",
    \\"contact_id\\": \\"2\\",
    \\"first_name\\": \\"Elon\\",
    \\"last_name\\": \\"Musk\\",
    \\"description\\": \\"A thinker\\",
    \\"prefix\\": \\"Sir\\",
    \\"title\\": \\"CEO\\",
    \\"status\\": \\"New\\",
    \\"monetary_amount\\": 75000,
    \\"currency\\": \\"USD\\",
    \\"fax\\": \\"+12129876543\\",
    \\"websites\\": [
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"addresses\\": [
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        },
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        }
    ],
    \\"social_links\\": [
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        },
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        }
    ],
    \\"phone_numbers\\": [
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"emails\\": [
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"custom_fields\\": [
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        },
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        }
    ],
    \\"tags\\": [
        \\"New\\"
    ]
}"
`;

exports[`Fuzzer should fuzz the required prop of the request body 1`] = `
"{
    \\"company_name\\": \\"Spacex\\",
    \\"owner_id\\": \\"54321\\",
    \\"company_id\\": \\"2\\",
    \\"contact_id\\": \\"2\\",
    \\"first_name\\": \\"Elon\\",
    \\"last_name\\": \\"Musk\\",
    \\"description\\": \\"A thinker\\",
    \\"prefix\\": \\"Sir\\",
    \\"title\\": \\"CEO\\",
    \\"status\\": \\"New\\",
    \\"monetary_amount\\": 75000,
    \\"currency\\": \\"USD\\",
    \\"fax\\": \\"+12129876543\\",
    \\"websites\\": [
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"addresses\\": [
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        },
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        }
    ],
    \\"social_links\\": [
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        },
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        }
    ],
    \\"phone_numbers\\": [
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"emails\\": [
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"custom_fields\\": [
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        },
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        }
    ],
    \\"tags\\": [
        \\"New\\"
    ]
}"
`;

exports[`Fuzzer should fuzz the value of a prop of the request body above the defined maximum 1`] = `
"{
    \\"name\\": \\"Elon Musk\\",
    \\"company_name\\": \\"Spacex\\",
    \\"owner_id\\": \\"54321\\",
    \\"company_id\\": \\"2\\",
    \\"contact_id\\": \\"2\\",
    \\"first_name\\": \\"Elon\\",
    \\"last_name\\": \\"Musk\\",
    \\"description\\": \\"A thinker\\",
    \\"prefix\\": \\"Sir\\",
    \\"title\\": \\"CEO\\",
    \\"status\\": \\"New\\",
    \\"monetary_amount\\": 301,
    \\"currency\\": \\"USD\\",
    \\"fax\\": \\"+12129876543\\",
    \\"websites\\": [
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"addresses\\": [
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        },
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        }
    ],
    \\"social_links\\": [
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        },
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        }
    ],
    \\"phone_numbers\\": [
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"emails\\": [
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"custom_fields\\": [
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        },
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        }
    ],
    \\"tags\\": [
        \\"New\\"
    ]
}"
`;

exports[`Fuzzer should fuzz the value of a prop of the request body below the defined minimum 1`] = `
"{
    \\"name\\": \\"Elon Musk\\",
    \\"company_name\\": \\"Spacex\\",
    \\"owner_id\\": \\"54321\\",
    \\"company_id\\": \\"2\\",
    \\"contact_id\\": \\"2\\",
    \\"first_name\\": \\"Elon\\",
    \\"last_name\\": \\"Musk\\",
    \\"description\\": \\"A thinker\\",
    \\"prefix\\": \\"Sir\\",
    \\"title\\": \\"CEO\\",
    \\"status\\": \\"New\\",
    \\"monetary_amount\\": 99,
    \\"currency\\": \\"USD\\",
    \\"fax\\": \\"+12129876543\\",
    \\"websites\\": [
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"url\\": \\"http://example.com\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"addresses\\": [
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        },
        {
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\",
            \\"name\\": \\"HQ US\\",
            \\"line1\\": \\"Main street\\",
            \\"line2\\": \\"apt #\\",
            \\"city\\": \\"San Francisco\\",
            \\"state\\": \\"CA\\",
            \\"postal_code\\": \\"94104\\",
            \\"country\\": \\"US\\",
            \\"latitude\\": \\"40.759211\\",
            \\"longitude\\": \\"-73.984638\\"
        }
    ],
    \\"social_links\\": [
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        },
        {
            \\"url\\": \\"https://www.twitter.com/apideck-io\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"twitter\\"
        }
    ],
    \\"phone_numbers\\": [
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"number\\": \\"111-111-1111\\",
            \\"id\\": \\"12345\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"emails\\": [
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        },
        {
            \\"email\\": \\"elon@musk.com\\",
            \\"id\\": \\"123\\",
            \\"type\\": \\"primary\\"
        }
    ],
    \\"custom_fields\\": [
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        },
        {
            \\"id\\": \\"custom_technologies\\",
            \\"value\\": \\"Uses Salesforce and Marketo\\"
        }
    ],
    \\"tags\\": [
        \\"New\\"
    ]
}"
`;
